### PR TITLE
Fix 18 codellama

### LINKED ISSUE #18: CodeLLaMa stop tokens sometimes breaking
There seems to be some issues with CodeLLaMa stop tokens that breaks it sometimes. I will try to get a working reproduction and understand what's going on.

CC: @claudiosv

### PR DESCRIPTION
**Pull request by DNGros**
Fixes #18

### CODE REVIEW

File: lmwrapper/HuggingfacePredictor.py [hunk 1]
```diff
@@ -525,3 +527,42 @@ def estimate_tokens_in_prompt(self, prompt: LmPrompt) -> int:
     @property
     def is_chat_model(self):
         return self._is_chat_model
+
+
+def _get_hf_token_offsets(tokenizer, token_ids) -> list[int]:
```
Comment (claudiosv) [inline lmwrapper/HuggingfacePredictor.py:532]: Can we type the arguments? Maybe something like:
```python
tokenizer: PreTrainedTokenizerFast, token_ids: torch.Tensor | list[int]
```

Also, not to be too nitpicky but maybe this could just be `_get_token_offsets` as it is already in the `HuggingfacePredictor` module?

File: lmwrapper/HuggingfacePredictor.py [hunk 2]
```diff
@@ -293,34 +293,36 @@ def new_call(*args, **kwargs):
...
             skip_special_tokens=True,
             clean_up_tokenization_spaces=True,
         )
+        token_offsets = _get_hf_token_offsets(self._tokenizer, generated_sequence)
+        token_offsets_full = []
+        for i in range(1, len(token_offsets)):
+            token_offsets_full.extend([i - 1] * (token_offsets[i] - token_offsets[i - 1]))
+
         if prompt.stop:
+            #if not "".join(generated_tokens) != generated_text:
```
Comment (claudiosv) [inline lmwrapper/HuggingfacePredictor.py:318]: Can we move this to a new method like `_verify_concatable`? For now it could do nothing but in the future we could add a flag for checking concatenability? If not can we remove the commented code?

File: lmwrapper/HuggingfacePredictor.py [hunk 3]
```diff
@@ -293,34 +293,36 @@ def new_call(*args, **kwargs):
...
+            #    msg = "Tokens do not appear to be concatenatable. This likely means the tokenizer is " \
+            #          "doing some extra processing or we need to better handle special tokens. For now " \
+            #          "stop sequences are unsupported."
+            #    raise NotImplementedError(msg)
             sorted_stop_sequences = sorted(prompt.stop, key=len, reverse=True)
 
-            stop_idx = len(generated_sequence)
+            #stop_idx = len(generated_sequence)
```
Comment (claudiosv) [inline lmwrapper/HuggingfacePredictor.py:325]: Nice catch, this does seem unnecessary. Can we delete this line?

File: lmwrapper/HuggingfacePredictor.py [hunk 4]
```diff
@@ -293,34 +293,36 @@ def new_call(*args, **kwargs):
 
         model_output_sequence = generation_output.sequences[
             0
-        ]  # we will not mutate this one
+        ].cpu()  # we will not mutate this one
```
Comment (claudiosv) [inline lmwrapper/HuggingfacePredictor.py:296]: Can we `detach()` this tensor as well? Detach returns a new tensor that is not associated with the graph. My understanding is that this is best practice, even though there are no gradients/autograd.

**Review by claudiosv — COMMENTED** [review]
Thanks for fixing this! I left some minor comments. No worries if you just want to merge.

File: lmwrapper/HuggingfacePredictor.py (hunk 4 above)
Comment (claudiosv) [reply lmwrapper/HuggingfacePredictor.py:296]: Update: A downside I noticed is that `model_output_sequence` is passed to `_gather_logprobs_from_logits` which applies softmax. This may have unintended performance impacts, though probably minor.

**Comment by claudiosv** [comment]
I committed the changes I suggested. Feel free to adjust.

File: lmwrapper/HuggingfacePredictor.py (hunk 4 above)
Comment (DNGros) [reply lmwrapper/HuggingfacePredictor.py]: good catch 👍

File: lmwrapper/HuggingfacePredictor.py (hunk 2 above)
Comment (DNGros) [reply lmwrapper/HuggingfacePredictor.py]: good idea. I don't think this actually needed now that using the tokenizer to get the offsets
